_source/_posts/2020-12-08-beginners-guide-to-jwt.adoc

```diff
@@ -0,0 +1,175 @@
+
+== JWT structure
+
+A JWS (the most common type of JWT) contains three parts separated by a dot (`.`). The first two parts (the "header" and "payload") are Base64-URL encoded JSON, and the third is a cryptographic signature.
+
+Let's look at an example JWT:
+
+[source,txt]
+----
+eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0.5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U
+----
+
+Breaking this down into the individual sections we have:
+
+[source,sh]
+----
+eyJhbGciOiJIUzI1NiJ9 # header
+.
+eyJuYW1lIjoiSm9lIENvZGVyIn0 # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: If you have a JWT with more than three sections, it's probably a JWE.
+
+Next, each of the first two sections are base64-url decoded:
+
+[source,json]
+----
+{"alg":"HS256"} # header
+.
+{"name":"Joe Coder"} # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: On macOS  (with "coreutils" installed, you can base64-url decode the strings on the command line with: `echo eyJhbGciOiJIUzI1NiJ9 | gbasenc -d --base64url`
+
+The last section in the JWT, the signature, is also base64-url encoded, but it's just binary data; if you try to decode it, you will end up with non-displayable characters:
+
+```txt
+��i�i�￹�￹KN�f￹�￹֪�O�_￹R��￹\�+￹
+```
+
+=== JWT Claims
+
+Once you start using JWTs you start hearing the word "claim" everywhere. A JWT claim is a key/value pair in a JSON object.  In the example above, `"name": "Joe Coder"`, the claim key is 'name' and the value is 'Joe Coder`. The value of a claim can be any JSON object.
+
+There are three types of claims: "registered," "public," and "private." You can find the list of registered and public claims in the https://www.iana.org/assignments/jwt/jwt.xhtml#claims[official IANA Registry].  You can also add any other custom claim to a JWT; these are known as "private claims."
+
+WARNING: When using private claims, watch out for name collisions with the official claims.
+
+The use of registered claims is optional, but when they are present, they MUST be validated.  For example, a JWT may contain date-time fields that describe when the token is valid.
+
+- Issued At (`iat`) - The time the JWT was created
+- Expiration Time (`exp`) - The time at which the JWT is no longer valid
+- Not Before (`nfb`) - The earliest time the JWT would be valid
+
+NOTE: Timestamps are "seconds since the epoch" integer format. JWT libraries can add up to a few minutes of leeway to these values to account for clock skew between systems.
+
+=== JWT Header
+
+The header of a JWT contains information about how the token was created.  In my example, the "algorithm" (`alg`) claim is set to `HS256`, which specifies the hashing algorithm HMAC SHA-256 is used to generate or validate the signature.
+
+[source,json]
+----
+{
+  "alg": "HS256"
+}
+----
+
```

Comment:
I'd add a strong warning here about `"alg": "NONE"` and why you should _avoid it like the plague_ and, when possible, configure your JWT library with an allow-list of expected algorithms.

Reply:
@jpf I think I hit most of your comments.  One notable exception is the location of the `none` algorithm warning.  My goal was to try to walk through the format, and then bring it up in the "Problems" section. That said, I DO want to make sure it obvious and is not missed.

I could also add short note in the next section about the "none" algorithm, and add a "See more about that in the 'Problems with JWT' below"?

Thoughts?

Reply:
I like your idea on adding a "See more about that ..." note. Do that!